tds/src/main/java/thredds/server/notebook/NotebookController.java

```diff
@@ -0,0 +1,309 @@
+package thredds.server.notebook;
+
+import org.json.*;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Controller;
+import org.springframework.validation.BindingResult;
+import org.springframework.web.bind.annotation.RequestMapping;
+import thredds.client.catalog.Catalog;
+import thredds.client.catalog.Dataset;
+import thredds.core.CatalogManager;
+import thredds.core.StandardService;
+import thredds.core.TdsRequestedDataset;
+import thredds.server.config.TdsContext;
+import thredds.util.Constants;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.validation.Valid;
+import java.io.*;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+
+@Controller
+@RequestMapping("/notebook")
+public class NotebookController {
+
+  private final String DS_REPLACE_TEXT = "{{datasetName}}";
+  private final String CAT_REPLACE_TEXT = "{{catUrl}}";
+
+  @Autowired
+  TdsContext tdsContext;
+
+  @Autowired
+  CatalogManager catalogManager;
+
+  // TODO: clean, refactor, use standard service bases
+  @RequestMapping("**")
+  public void getNotebookForDataset(HttpServletRequest req, HttpServletResponse res, @Valid NotebookParamsBean params,
+                                    BindingResult validationResult) throws Exception {
+
+    // Get Catalog and Dataset
+    TdsRequestedDataset reqD = new TdsRequestedDataset(req, getBase());
+    String datasetId = reqD.getPath();
+
+    String catalogName = params.catalog;
+    if (catalogName == null) {
+      throw new IllegalArgumentException("Argument 'catalog' cannot be null.");
+    }
+
+    Catalog catalog = getCatalog(catalogName, req);
+    Dataset dataset = catalog.findDatasetByID(datasetId);
+
+    // Get notebook
+    File responseFile = getNotebookFile(dataset, catalog);
+    if (responseFile == null) {
+      // TODO: handle

```

Comment:
There are a bunch of "TODO" comments concerning warning/logging in this controller - I'm not sure how we generally handle warnings, but there are situations here, such as missing fields in the ipynb metadata, which should probably just warn and return a default rather than throwing an error.

Reply:
I think I answered this in email, but to capture it here, basically follow `thredds.server.exception.ServiceNotAllowed` and ` thredds.server.TdsErrorHandling` to get messages to show up in `threddsServlet.log`